snapshot.go

```diff
@@ -406,43 +404,30 @@ func WriteSnapshot(ctx context.Context, tx uint64, db *DB, w io.Writer, offline
 				},
 			}
 
-			var ascendErr error
-			block.Index().Iterate(func(node *index.Node) bool {
+			if err := block.Index().Snapshot(func(p parts.Part) error {
 				granuleMeta := &snapshotpb.Granule{}
-				if node.Part() == nil {
-					return true
-				}
-				p := node.Part()
 				partMeta := &snapshotpb.Part{
 					StartOffset:     int64(offW.offset),
 					Tx:              p.TX(),
 					CompactionLevel: uint64(p.CompactionLevel()),
 				}
-				err := func() error {
-					if err := ctx.Err(); err != nil {
+				if err := ctx.Err(); err != nil {
+					return err
+				}
+				schema := t.schema
+
+				if record := p.Record(); record != nil {
```

Comment:
One other thing I would do to clean this up is maybe adding a `Write(w io.Writer)` to the part interface to avoid having the snapshot code figure this out.

Reply:
oh I like that idea!

Reply:
done